### PR TITLE
Feature/2019 09 12 api first development with open api or swagger

### PR DESCRIPTION
**Pull request by Nxtra**

### CODE REVIEW

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+We also see the following trends in software engineering:
+* **Shift towards the cloud.**  
+Instead of big monolitic applications we are building lot's of smaller (micro)services.
+All communication between those services goes through the API.
+
+* **Multiple frontend applications use the same backend.**  
+Often these applications are created by separate teams.
+
+* **API carry business value.**  
+Yes, there is money in your API.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: use double space to begin a new line

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+An API exposes the functionality of your product.
+A good API allows user to integrate with your product with ease.
+Thus making your product a great choice.
+
+In all of the cases a above there is value (even money!) in good communication between services.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: scratch the (even money!)

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+And that's why you should practice API first development.
+Put your communication first!
+
+> API first development makes a clear API definition the first priority
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: The first priority in your API first development story is a clear API definition

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+* I'll start of with a lecture about API first development and it's advantages.([Jump to section](#api-first-development---why-how-and-what))
+* We will discover how to visualize our API specs. ([Jump to section](#api-first-development-with-openapiswagger))  
+* Generate a postman collection from your OpenAPI definition. ([Jump to section](#integrating-with-postman))   
+* Next I'll show you how to generate code that is completely compliant with your API specs. ([Jump to section](#openapi-generator-generate-api-compliant-code))
+* We then dive into how to integrate this in your build process by using maven or gradle plugins. ([Jump to section](#integrating-with-your-build-process-maven-or-gradle-plugin))
+* To finish it up I'll demonstrate how to use it in a cloud native serverless product with `AWS SAM` and `AWS Lambda`. ([Jump to section](#serverless-on-aws-openapi-api-gateway-lambda-and-sam))
+* Wrapping it up with a conclusion.
+
+# API first development - Why, how and what
+> Great communication is key to great software engineering.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: A little intro about the concept of API economy would be nice. 
For what stands the abbreviation.

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+
+That also goes for applications and services.
+If your applications cannot communicate properly, you'll never be able to expose the functionality that is key to a good product.
+
+
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Good list, you can add stateless services

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+Thus making your product a great choice.
+
+In all of the cases a above there is value (even money!) in good communication between services.
+And that's why you should practice API first development.
+Put your communication first!
+
+> API first development makes a clear API definition the first priority
+
+***How can I practice API first development?***
+* Understand that the API is the interface for your application.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Rephrase it in to more of story. 
for ex. 
Defining APIs gives us the opportunity to expose our application's functionality and create bridges between our provider and consumer.

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+In all of the cases a above there is value (even money!) in good communication between services.
+And that's why you should practice API first development.
+Put your communication first!
+
+> API first development makes a clear API definition the first priority
+
+***How can I practice API first development?***
+* Understand that the API is the interface for your application.  
+It is the intersection where multiple services join hands to couple there functionality.
+* The first step is to design your API.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Use more comma's, you are hacking your sentences now and it doesn't read fluently.

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+The implementation comes next.
+This will allow teams to develop their applications separately because they both know and understand how communication between the services will happen.
+The contract between services is set.
+
+> API first development allows teams to develop separately against a common interface, the API.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Leave 'the API' out

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+
+Now that we understand the importance of and value of API first design let's see how the Swagger/OpenAPI spec can help you with that.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
Now that we understand the importance of and value of API first design, let's see how the Swagger/OpenAPI spec can help you with that.
```

Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Use we or you format, not both

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+Users can check session information and register for sessions.
+The backend service is accessed by a web application and two mobile apps.
+
+<div style="text-align: center;">
+  <img src="/img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/design.png" width="40%" height="40%" target="_blank">
+</div>
+
+The applications are created by different teams and they all embrace the **API first** approach.
+
+They read this blog and realised that by agreeing on the common interface first they could develop separately without impacting each other.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
They read this blog and realised that by agreeing on the common interface first, they could develop separately without impacting each other.
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+* A client should be able to create a new session via the API
+
+This is a crucial part of the API first mindset.
+We need to clearly define and communicate the API before starting to implement.
+Designing an API is easier when you can visualise the API.
+Let's bring in the OpenAPI spec.
+
+> The OpenAPI specification allows you to define your API in a descriptive language (JSON or Yaml) and nicely visualise it
+
+Note: If I speak about OAS (OpenAPI Specification) instead of OpenAPI/Swagger specification.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Rephrase what you mean here

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+We need to clearly define and communicate the API before starting to implement.
+Designing an API is easier when you can visualise the API.
+Let's bring in the OpenAPI spec.
+
+> The OpenAPI specification allows you to define your API in a descriptive language (JSON or Yaml) and nicely visualise it
+
+Note: If I speak about OAS (OpenAPI Specification) instead of OpenAPI/Swagger specification.  
+Let's now use OAS to help us with our API First approach and design our API.
+
+Let me introduce you to [https://editor.swagger.io](https://editor.swagger.io){:target="_blank"}.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
Let me introduce you to [https://editor.swagger.io](https://editor.swagger.io){:target="_blank"}, a portal to visualise...
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+A portal to visualise your OAS.
+Easy to use and offering all the functionality we need for this example.
+
+I'll keep it simple and create the OAS for the API exposing the endpoint where consumers of the API can fetch all sessions of the conference.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: I'll keep it simple, we will create the OAS for exposing the endpoint to let consumers fetch the sessions of the conference

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+
+The API in the example is small.
+When describing a whole real-world API, the file might become quite large.
+But that's no problem.
+The OpenAPI spec allows you to split your definitions over multiple files which you can reference from within other files.
+
+> Takeaways: Easy descriptive language & great visualisations
+
+You want to expose your beautiful visualisation to your clients.
+They shouldn't have to past a `yaml` file in a window of there browser all the time.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
They shouldn't have to paste a `yaml` file in a window of there browser all the time.
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 15]
```diff
@@ -0,0 +1,549 @@
...
+```
+If this finds any violations throw a `ConstraintValidationException`.
+
+Again a unit test can validate that:
+* no violations are found when the response is validated
+* no exception is thrown
+
+
+# Serverless on AWS: OpenAPI, API Gateway Lambda and SAM
+It's fairly easy to create an API Gateway from a openAPI specification.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
It's fairly easy to create an API Gateway from an openAPI specification.
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+In the API Gateway console under *Create* select *Import from Swagger or Open API 3* and upload your specification.
+
+<div style="text-align: center;">
+  <img src="/img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/create-apigateway-via-ui.gif" width="100%" height="100%">
+</div>
+
+Of course we want to use the specification programmatically.
+Suppose we create our backend service with AWS Lambda (serverless).
+I'll keep to AWS native tools and will use SAM to deploy the Lambda functions and my API.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
I'll be using AWS native tools and use SAM to deploy the Lambda functions and my API.
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 17]
```diff
@@ -0,0 +1,549 @@
...
+SAM allows you to use an OpenAPI specification to create your API Gateway.
+
+In your SAM template define the API Gateway resource by referencing you OAS.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
In your SAM template define the API Gateway resource by referencing your OAS.
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 18]
```diff
@@ -0,0 +1,549 @@
...
+      Events:
+        CreateSessionApi:
+          Type: Api
+          Properties:
+            RestApiId: !Ref "ConferenceApiGateway"
+            Path: /sessions
+            Method: GET
+```
+
+And add the `x-amazon-apigateway-integration` extension in your `api.yml` to specify how the api has to integrate with the backend Lambda service.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion 
Add the `x-amazon-apigateway-integration` extension in your `api.yml` to specify how the api has to integrate with the backend Lambda service.
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 19]
```diff
@@ -0,0 +1,549 @@
...
+    get:
+      ...
+      x-amazon-apigateway-integration:
+        type: "aws_proxy"
+        httpMethod: "POST"
+        uri:
+          Fn::Sub: arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${GetAllSessionsFunction.Arn}/invocations
+```
+
+For a the full template including the Lambda resources look in this [gist](https://gist.github.com/Nxtra/3600ccab85d92faebf3c465701ba7c21){:target="_blank"}.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
For a full template example including the Lambda resources, check out this [gist](https://gist.github.com/Nxtra/3600ccab85d92faebf3c465701ba7c21){:target="_blank"}.
```

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 20]
```diff
@@ -0,0 +1,549 @@
...
+
+    <dependency>
+        <groupId>io.springfox</groupId>
+        <artifactId>springfox-bean-validators</artifactId>
+        <version>${springfox.version}</version>
+    </dependency>
+</dependencies>
+```
+
+Implementing the API and using the write annotations leads to an endpoint of your application on which your API spec is visualised: `/swagger-ui.html`
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: write or right?

File: img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/api.yml [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+    post:
+      description: Create a new session on the conference
+      operationId: createSession
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/Session"
+      responses:
+        '202':
```
Comment (KevinVHoutte) [inline img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/api.yml]: POST requests return created 201 status code with a location header

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md (hunk 5 above)
Comment (Nxtra) [reply _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: How would I relate that to the need of having a good API?

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md (hunk 6 above)
Comment (Nxtra) [reply _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: I like the way you phrased it. 
Which part do you want me to rewrite exactly?

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 22]
```diff
@@ -0,0 +1,549 @@
...
+If you followed along clear the directory with
+```bash
+find . ! -name api.yml -delete
+```
+
+For starters we only want to generate the model classes:
+```bash
+openapi-generator generate -i api.yml -g java -Dmodels
+```
+Models here refers to your DTO's (Data Transfer Objects).
```
[the remark this answers is not in the document]
Comment (Nxtra) [reply _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Voorstel: Models here refers to the resources of your API. In the code we'll typically call them DTO's (Data Transfer Objects)

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md (hunk 20 above)
Comment (Nxtra) [reply _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: yeah right :D

File: img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/api.yml (hunk 21 above)
Comment (Nxtra) [reply img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/api.yml]: Ik verwerk die create async. Vandaar de 202 .

Comment (KevinVHoutte) [reply img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/api.yml]: Ok dan heb ik niks gezegd ;)

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md (hunk 5 above)
Comment (KevinVHoutte) [reply _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]: Maybe it's overkill nvm it

File: img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/api.yml (hunk 21 above)
Comment (Nxtra) [reply img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/api.yml]: Ik zal dat vermelden. Veel mensen zullen anders die reacties hebben.

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (KevinVHoutte) [reply _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
Models here refers to your DTO's (Data Transfer Objects) or Resources.
```
:D

File: _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md [hunk 23]
```diff
@@ -0,0 +1,549 @@
...
+* **Multiple frontend applications use the same backend.**  
+Often these applications are created by separate teams.
+
+* **API carry business value.**    
+Yes, there is money in your API.  
+An API exposes the functionality of your product.
+A good API allows user to integrate with your product with ease.
+Thus making your product a great choice.
+
+In all of the cases a above there is value in good communication between services.
```
Comment (KevinVHoutte) [inline _posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md]:
```suggestion
In all of the cases above, there is value in good communication between services.
```

File: docker-compose.yml [hunk 24]
```diff
@@ -5,7 +5,7 @@ services:
     build:
       dockerfile: Dockerfile.build
       context: .
-    command: jekyll serve --incremental
+    command: jekyll serve --incremental --future
```
Comment (KevinVHoutte) [inline docker-compose.yml]: remove it back

**Pull request merged**